File: Production/Source/Ripley/DateRangeValidation.cpp
/*

Copyright British Broadcasting Corporation 2001.

This code is owned by the BBC and must not be copied, 
reproduced, reconfigured, reverse engineered, modified, 
amended, or used in any other way, including without 
limitation, duplication of the live service, without 
express prior written permission of the BBC.

The code is provided "as is" and without any express or 
implied warranties of any kind including without limitation 
as to its fitness for a particular purpose, non-infringement, 
compatibility, security or accuracy. Furthermore the BBC does 
not warrant that the code is error-free or bug-free.

All information, data etc relating to the code is confidential 
information to the BBC and as such must not be disclosed to 
any other party whatsoever.

*/

#include "stdafx.h"
#include "DateRangeValidation.h"

#ifdef _DEBUG
#undef THIS_FILE
static char THIS_FILE[]=__FILE__;
#define new DEBUG_NEW
#endif

CDateRangeValidation::CDateRangeValidation(CInputContext& inputContext) : CXMLObject(inputContext), m_TimeInterval(0)
{
	m_LastValidationRes = CDateRangeValidation::NOTVALIDATEDYET;
}

CDateRangeValidation::~CDateRangeValidation(void)
{
}


/*********************************************************************************

	ValidationResult CDateRangeValidation::ValidateDateRange(CTDVDateTime StartDate, CTDVDateTime EndDate, int TimeInterval = 1, bool ErrorOnFutureDates = false, bool ErrorOnInvalidTimeInterval = true, bool bCreateEndDate = true, bool bValidateTimeInterval = false)

		Author:		Mark Howitt
		Created:	18/01/2007
		Inputs:		CTDVDateTime StartDate - The start date for the range
					CTDVDateTime EndDate - The end date for the range
					int TimeInterval - An interval of time within the range.
						e.g. A day in the month of may = 1, A week in the month of may = 7.
					bool ErrorOnFutreDates - A flag that states that you want to force an error on futre dates
					bool ErrorOnInvalidTimeInterval - A flag that states whether you want to error on invalid time intervals
					bool bCreateEndDate - A flag to say that we are needed to create the end date and make it a day later than the given start date
		Outputs:	-
		Returns:	The validation result
		Purpose:	Checks to make sure the supplied date range is valid

*********************************************************************************/
CDateRangeValidation::ValidationResult CDateRangeValidation::ValidateDateRange(CTDVDateTime StartDate, CTDVDateTime EndDate, int TimeInterval, bool ErrorOnFutureDates, bool ErrorOnInvalidTimeInterval, bool bValidateTimeInterval)
{
	// Set the member variables to the supplied values;
	m_StartDate = StartDate;
	m_EndDate = EndDate;
	m_TimeInterval = TimeInterval;

	// Check to make sure the start date is valid
	if (!m_StartDate.GetStatus())
	{
		SetDNALastError("CDateRangeValidation::ValidateDateRange","StartDateInvalid","Start date is invalid");
		return SetLastValidationResult(STARTDATE_INVALID);
	}

	// Check to make sure the end date is valid
	if (!m_EndDate.GetStatus())
	{
		SetDNALastError("CDateRangeValidation::ValidateDateRange","EndDateInvalid","End date is invalid");
		return SetLastValidationResult(ENDDATE_INVALID);
	}

	// Check to see if the dates are within valid datetime ranges for the database. 
    if (!m_StartDate.IsWithinDBSmallDateTimeRange())
    {
		// Invalid information given
        SetDNALastError("CDateRangeValidation::ValidateDateRange","StartDateInvalid","Start date is invalid");
		return SetLastValidationResult(STARTDATE_INVALID);
    }

	if (!m_EndDate.IsWithinDBSmallDateTimeRange())
	{
		SetDNALastError("CDateRangeValidation::ValidateDateRange","EndDateInvalid","End date is invalid");
		return SetLastValidationResult(ENDDATE_INVALID);
	}

	// Now check to see if we're worried about fututre dates
	if (ErrorOnFutureDates)
	{
		// Check to see if the start date is in the future
		CTDVDateTime CurrentDate(COleDateTime::GetCurrentTime());
		if (m_StartDate > CurrentDate)
		{
			SetDNALastError("CDateRangeValidation::ValidateDateRange","StartDateInTheFuture","Start date is in the future");
			return SetLastValidationResult(FUTURE_STARTDATE);
		}

		// Now check to see if the end date is in the future
		if (m_EndDate > CurrentDate)
		{
			SetDNALastError("CDateRangeValidation::ValidateDateRange","EndDateInTheFuture","End date is in the future");
			return SetLastValidationResult(FUTURE_ENDDATE);
		}
	}

	// Check tyo make sure the start date is not the same as the end date
	if (m_StartDate == m_EndDate)
	{
		SetDNALastError("CDateRangeValidation::ValidateDateRange","StartDateEqualsEndDate","Start date equals end date");
		return SetLastValidationResult(STARTDATE_EQUALS_ENDDATE);
	}

	// Check tomake sure that the start date is before the end date
	if (m_StartDate > m_EndDate)
	{
		SetDNALastError("CDateRangeValidation::ValidateDateRange","StartDateGreaterThanEndDate","Start date greater than end date");
		return SetLastValidationResult(STARTDATE_GREATERTHAN_ENDDATE);
	}

	// If dealing with fuzzy dates
	if (m_TimeInterval < 1 && bValidateTimeInterval)
	{ 
		SetDNALastError("CDateRangeValidation::ValidateDateRange","Timeinterval invalid","The timeinterval is less than 1.");
		return SetLastValidationResult(TIMEINTERVAL_INVALID);
	}

	// Check to see if we want invalid time intervals as errors
	if (ErrorOnInvalidTimeInterval)
	{
		// Check to make sure the interval is smaller or equal to the difference in dates
		int ElapsedDaysDiff = (int)(m_EndDate.m_dt - m_StartDate.m_dt);
		if (ElapsedDaysDiff < TimeInterval)
		{
			SetDNALastError("CDateRangeValidation::ValidateDateRange","TimeIntervalInvalid","Time interval invalid for date range");
			return SetLastValidationResult(TIMEINTERVAL_INVALID);
		}
	}

	// Make sure the last dna error gets removed.
	ClearError();

	// Got here? valid date range and interval
	return SetLastValidationResult(VALID);
}

/*********************************************************************************

	CDateRangeValidation::ValidationResult CDateRangeValidation::SetLastValidationResult(CDateRangeValidation::ValidationResult res)

		Author:		Mark Neves
		Created:	12/02/2007
		Inputs:		A result code
		Outputs:	-
		Returns:	The same code that's passed in
		Purpose:	Helper func for storing the last validation result the class has made

*********************************************************************************/

CDateRangeValidation::ValidationResult CDateRangeValidation::SetLastValidationResult(CDateRangeValidation::ValidationResult res)
{
	m_LastValidationRes = res;
	return m_LastValidationRes;
}

File: Production/Source/Shared/BBC.Dna.Utils/DateRangeValidation.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace BBC.Dna.Utils
{
    /// <summary>
    /// DateRangeValidation - A helper calss that validates date ranges
    /// </summary>
    public class DateRangeValidation
    {
        /// <summary>
        /// Enum of possible validation results
        /// </summary>
        public enum ValidationResult
        {
            /// <summary>
            /// Value for a valid validation test
            /// </summary>
		    VALID,
            /// <summary>
            /// Value for an invalid start date
            /// </summary>
		    STARTDATE_INVALID,
            /// <summary>
            /// Value for an invalid end date
            /// </summary>
		    ENDDATE_INVALID,
            /// <summary>
            /// Value for a start date that is equal to the end date
            /// </summary>
		    STARTDATE_EQUALS_ENDDATE,
            /// <summary>
            /// Value for a start date that is greater than the end date
            /// </summary>
		    STARTDATE_GREATERTHAN_ENDDATE,
            /// <summary>
            /// Value for an invalid time interval
            /// </summary>
		    TIMEINTERVAL_INVALID,
            /// <summary>
            /// Value for a start date that is in the future
            /// </summary>
		    FUTURE_STARTDATE,
            /// <summary>
            /// Value for an end date in the future
            /// </summary>
		    FUTURE_ENDDATE
	    };

        // The last validation result
        private ValidationResult _lastValidationResult = ValidationResult.VALID;
        private DateTime _startDate;
        private DateTime _endDate;

        /// <summary>
        /// Last result property
        /// </summary>
        public ValidationResult LastResult
        {
            get { return _lastValidationResult; }
            set { _lastValidationResult = value; }
        }

        /// <summary>
        /// Last start date parsed property
        /// </summary>
        public DateTime LastStartDate
        {
            get { return _startDate; }
        }

        /// <summary>
        /// Last end date parsed property
        /// </summary>
        public DateTime LastEndDate
        {
            get { return _endDate; }
        }

       /// <summary>
        /// Validate a date to see if it represents a valid date . 
        /// </summary>
        /// <param name="date">The start date for the date range</param>
        /// <param name="errorOnFutureDate">A flag that states whether or not you want a date in the future to be error conditions</param>
        /// <returns>The result of the validation</returns>
        public ValidationResult ValidateDate(DateTime date, bool errorOnFutureDate)
        {
            // Reset the last result
            _lastValidationResult = ValidationResult.VALID;

            // Check to see if the dates are within valid datetime ranges for the database. 
            if (!DnaDateTime.IsDateWithinDBSmallDateTimeRange(date))
            {
                _lastValidationResult = ValidationResult.STARTDATE_INVALID;
                return _lastValidationResult;
            }
	        // Now check to see if we're worried about future dates
            if (errorOnFutureDate)
            {
                // Check to see if the start date is in the future
                if (date > DateTime.Today)
                {
                    // Start date in the future
                    _lastValidationResult = ValidationResult.FUTURE_STARTDATE;
                    return _lastValidationResult;
                }
            }
            // Got here? Must be valid!
            return ValidationResult.VALID;
        }
        /// <summary>
        /// Validate two dates to see if they represent a valid date range. 
        /// </summary>
        /// <param name="startDate">The start date for the date range</param>
        /// <param name="endDate">The end date for the date range</param>
        /// <param name="timeInterval">A time interval within the date range. e.g. 1 = a day, 7 = a week</param>
        /// <param name="errorOnFutureDates">A flag that states whether or not you want dates in the future to be error conditions</param>
        /// <param name="errorOnInvalidTimeInterval">A flag that states whether or not you want invalid time intervals to be error conditions</param>
        /// <returns>The result of the validation</returns>
        public ValidationResult ValidateDateRange(DateTime startDate, DateTime endDate, int timeInterval, bool errorOnFutureDates, bool errorOnInvalidTimeInterval)
        {
            // Reset the last result
            _lastValidationResult = ValidationResult.VALID;

            // Update the member dates
            _startDate = startDate;
            _endDate = endDate;

           	// Check to see if the dates are within valid datetime ranges for the database. 
	        if (!DnaDateTime.IsDateWithinDBSmallDateTimeRange(startDate))
	        {
                _lastValidationResult = ValidationResult.STARTDATE_INVALID;
                return _lastValidationResult;
	        }

            if (!DnaDateTime.IsDateWithinDBSmallDateTimeRange(endDate))
	        {
                _lastValidationResult = ValidationResult.ENDDATE_INVALID;
                return _lastValidationResult;
            }

	        // Now check to see if we're worried about future dates
	        if (errorOnFutureDates)
	        {
		        // Check to see if the start date is in the future
		        if (startDate > DateTime.Today)
		        {
			        // Start date in the future
			        _lastValidationResult = ValidationResult.FUTURE_STARTDATE;
                    return _lastValidationResult;
		        }

		        // Now check to see if the end date is in the future
		        if (endDate > DateTime.Today)
		        {
                    // End date is in the future
			        _lastValidationResult = ValidationResult.FUTURE_ENDDATE;
                    return _lastValidationResult;
		        }
	        }

	        // Check tyo make sure the start date is not the same as the end date
	        if (startDate == endDate)
	        {
                // Start date is equal to the end date.
		        _lastValidationResult = ValidationResult.STARTDATE_EQUALS_ENDDATE;
                return _lastValidationResult;
	        }

	        // Check tomake sure that the start date is before the end date
	        if (startDate > endDate)
	        {
                // The start date is greater than the end date
		        _lastValidationResult = ValidationResult.STARTDATE_GREATERTHAN_ENDDATE;
                return _lastValidationResult;
	        }

            // See if we're checking for invalid time intervals
            if (errorOnInvalidTimeInterval)
            {
                // Check to make sure the interval is smaller or equal to the difference in dates
                TimeSpan dateDiff = endDate.Subtract(startDate);
                if (dateDiff.Days < timeInterval)
                {
                    _lastValidationResult = ValidationResult.TIMEINTERVAL_INVALID;
                    return _lastValidationResult;
                }
            }

            // Got here? Must be valid!
            return ValidationResult.VALID;
        }

        /// <summary>
        /// Gets the last validtion result as an Error XML Element.
        /// </summary>
        /// <param name="doc">The document the element is to be added to.</param>
        /// <returns>An XML Element that represents the last validation result as an error.</returns>
        public XmlElement GetLastValidationResultAsXmlElement(XmlDocument doc)
        {
            // Create the new element and add the default attribute
            XmlElement validationResult = doc.CreateElement("ERROR");
            validationResult.SetAttribute("OBJECT", "DateRangeValidation.ValidateDateRange()");

            XmlText errorMsg = doc.CreateTextNode("");
            validationResult.AppendChild(errorMsg);

            // Now switch on the last result
            switch (_lastValidationResult)
            {
                case ValidationResult.VALID:
                    {
                        validationResult.SetAttribute("CODE","NoError");
                        errorMsg.Value = "No Error";
                        break;
                    }
                case ValidationResult.STARTDATE_INVALID:
                    {
                        validationResult.SetAttribute("CODE", "StartDateInvalid");
                        errorMsg.Value = "Start date is invalid";
                        break;
                    }
                case ValidationResult.ENDDATE_INVALID:
                    {
                        validationResult.SetAttribute("CODE", "EndDateInvalid");
                        errorMsg.Value = "End date is invalid";
                        break;
                    }
                case ValidationResult.STARTDATE_GREATERTHAN_ENDDATE:
                    {
                        validationResult.SetAttribute("CODE", "StartDateGreaterThanEndDate");
                        errorMsg.Value = "Start date greater than end date";
                        break;
                    }
                case ValidationResult.STARTDATE_EQUALS_ENDDATE:
                    {
                        validationResult.SetAttribute("CODE", "StartDateEqualsEndDate");
                        errorMsg.Value = "Start date equals end date";
                        break;
                    }
                case ValidationResult.TIMEINTERVAL_INVALID:
                    {
                        validationResult.SetAttribute("CODE", "TimeIntervalInvalid");
                        errorMsg.Value = "Time interval invalid for date range";
                        break;
                    }
                case ValidationResult.FUTURE_STARTDATE:
                    {
                        validationResult.SetAttribute("CODE", "StartDateInFuture");
                        errorMsg.Value = "Start date is in the future";
                        break;
                    }
                case ValidationResult.FUTURE_ENDDATE:
                    {
                        validationResult.SetAttribute("CODE", "EndDateInTheFuture");
                        errorMsg.Value = "End date is in the future";
                        break;
                    }
            }

            // Return the new element
            return validationResult;
        }
    }
}
